### PR TITLE
DOC Add intelex inference example

### PR DESCRIPTION
**Pull request by ahuber21**
I want to give a self-contained example how scikit-learn-intelex speeds up model inference times. This effort was kicked off in #251 and the PR in this form is basically just to align the requirements.

I'm new to the discussion, so I'm including just the most basic things. Please let me know if this example is a step in the right direction.

Output from running on my machine:
```
$ python examples/use_intelex.py
Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)
[skl] Inference took t_stock = 3.7e+00s and achieved 21.7% accuracy
[skl-ex] Inference took t_opt = 2.3e-01s and achieved 21.7% accuracy
t_stock / t_opt = 16.1
# ... more output about pushing and downloading ...
[skl] Inference took t_stock = 3.8e+00s and achieved 21.7% accuracy
[skl-ex] Inference took t_opt = 2.5e-01s and achieved 21.7% accuracy
t_stock / t_opt = 15.3
Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)
[skl] Inference took t_stock = 3.8e+00s and achieved 21.7% accuracy
```

### CODE REVIEW

File: examples/use_intelex.py
Comment (BenjaminBossan) [inline examples/use_intelex.py:7]: :class:`KNeighborsClassifier`

should probably be

    :class:`sklearn.neighbors.KNeighborsClassifier`

right now sphinx doesn't link it.

File: examples/use_intelex.py
Comment (BenjaminBossan) [inline examples/use_intelex.py:206]:
```suggestion
# were loaded with `patch_sklearn()`. Note, however, that this will not result
```

File: examples/use_intelex.py
Comment (BenjaminBossan) [inline examples/use_intelex.py:6]: Maybe if there is a link to a document showing what models are supported, it would be nice to link it here.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for adding this example. This is not an in depth review yet, just what I found on first look:

The example is not actually being run but just included as is into the docs, as can be seen here: 

https://skops--303.org.readthedocs.build/en/303/auto_examples/use_intelex.html#sphx-glr-auto-examples-use-intelex-py

I'm not completely sure why that happens here but not with the other examples, Possibly, this example needs to be linked explicitly in the docs somewhere? A good place would be in:

https://github.com/skops-dev/skops/blob/3e1f138c5a7c12863ca263beb2804ceede07d35a/docs/examples.rst

(which would be a good addition regardless of whether that is the root of the issue or not)

In general, I think this guide could be more useful if it was made a bit more accessible. E.g. explain in a few words what intelex does, link to its docs, and avoid too many abbreviations like "skl" and "sklex".

**Comment by ahuber21** [comment]
Hi @BenjaminBossan 
thanks for the feedback! I'm happy to hear that the original suggestions was already a step in the right direction.
I added your requested changes and also added some content to the introduction in fixup commits that we can squash later.
Let's see if the code actually runs this time. 

Let me know what else you want to see!

**Comment by BenjaminBossan** [comment]
The example is still not building correctly but I believe I have figured out the issue. It seems like the file name should always start with `plot_`. Could you please try if that fixes the issue?

Btw. you can build the docs locally by going into `docs/` and running `make html`, then open the html files in `_build/html`.

**Comment by ahuber21** [comment]
Hi @BenjaminBossan thanks for the tip. The latest commit worked locally so I hope this time the example runs.

I had an interesting thought. At the bottom of my example I show that loading a pickled non-optimized model has no speedup, even when `sklearnex` is loaded. I'm curious if that is also true for `skops.io` methods. Essentially, if a new object is created, i.e. the constructor is called, Intel optimizations should apply. It would be interesting to see if there are any compatibility issues.
I think @adrinjalali already looked into that, maybe he can comment.

**Comment by BenjaminBossan** [comment]
> The latest commit worked locally so I hope this time the example runs.

The doc build is failing, but it's still progress :) The trouble now is that building the docs has a dependency on sklearnex, which is not installed. Could you please add it to the [dependencies](https://github.com/skops-dev/skops/blob/main/skops/_min_dependencies.py) as a "doc" dependency?

> I'm curious if that is also true for `skops.io` methods. Essentially, if a new object is created, i.e. the constructor is called, Intel optimizations should apply.

I think this will depend on how exactly the patching works, which I don't know. Ideally, there would be an easy way to check on each estimator if it was patched, something like `hasattr(estimator, "_uses_intelex")` or `sklearnex.check_is_patched(estimator)`, I believe this would be really useful.

**Comment by ahuber21** [comment]
Progress indeed. The `sklearnex.check_is_patched(estimator)` suggestion is very good. I'm wondering if we already have something like this. Let me ping @napetrov, maybe he knows more.

**Comment by ahuber21** [comment]
By the way, the pipeline now complains about the missing HF token. I followed the approach that I found in `plot_hf_hub.py`, so I think the error is nothing new. How is this expected to work?

File: examples/plot_intelex.py
Comment (BenjaminBossan) [inline examples/plot_intelex.py:18]: Is there any check in sklearnex that a patched model produces the exact same output? So e.g. can I assume that `np.allclose(clf.predict_proba(X), clf_opt.predict_proba(X))`? Or is it more like the algorithm is the same but results can be different, e.g. due to random initialization? If it's the former, it would be nice to mention that.

File: examples/plot_intelex.py
Comment (BenjaminBossan) [inline examples/plot_intelex.py:108]: Let's rename `prediction` to `y_proba`, which is the canonical name for the output from `predict_proba`. The same for variables further below.

File: examples/plot_intelex.py
Comment (BenjaminBossan) [inline examples/plot_intelex.py:111]: Calling this "accuracy" is very confusing. I was surprised to find an "accuracy" of 21% on a binary classification task :D Let's either call it "log_loss_score" or something like that, or calculate the actual accuracy and not log loss.

File: examples/plot_intelex.py
Comment (BenjaminBossan) [inline examples/plot_intelex.py:131]: Finding the "same result" doesn't really tell us much here IMHO. This is KNN with 3 neighbors, so each probability is either 0.0, 0.33, 0.67, or 1.0. Since there are only 4 discrete values, it's not too surprising to get the exact same output. The argument would be stronger with something like a linear regression.

In the end, however, this comes back to what I remarked earlier, whether we can expect the exact same output after patching or not. Clarifying this would help a lot.

File: examples/plot_intelex.py
Comment (BenjaminBossan) [inline examples/plot_intelex.py:137]: It could be worth mentioning that sklearnex will be used on the Hub, e.g. for the Inference API. This might not be obvious to everyone.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks Andreas for adding this example. I have now given it a more thorough review. Overall it looks good, but I think a few things could be improved further. Please take a look at my comments.

> By the way, the pipeline now complains about the missing HF token. I followed the approach that I found in plot_hf_hub.py, so I think the error is nothing new. How is this expected to work?

Yes, that's unfortunately expected, see #47. As soon as this is merged, it should work, as it would use the HF token set up for this repo. When I tested this locally with my token, it worked.

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:20]: I basically copied a paragraph from an older Medium blog post. In the updated version it is replaced by a more high level description and links to two Medium blog posts (for anyone who needs more details). I first wasn't sure if I'm supposed to link to Medium in this example, but it might just be the best solution.

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:18]: sklearnex's backend is completely separate from scikit-learn, including a separate RNG. So there's no guarantee that identical outputs are achieved and, in fact, they're mostly different.
The agreement of model outputs is however tested in our CI https://github.com/intel/scikit-learn-intelex/blob/master/tests/test_examples.py

I'll mention this in the example

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:131]: Hmm you're right. At some point while condensing the initial example down to something quick and easy this part lost all its meaning. If you would like to see something else I'm happy to make the change. I mean, it's at least showing that (a) there is a result and it actually works, and (b) it's not totally buggy. That, in combination with the comment I'll add in the next commit may be enough. Let me know what you think

**Comment by ahuber21** [comment]
Hi Benjamin. Thanks! I hope all your comments are addressed with the latest commit.

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:66]: we'd need to see how long it takes on the CI machines, might need to reduce the number if it's too slow.

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:99]: I think in a script, this is very confusing. It'd be nicer if we explicitly import the class from sklearnex for it to be clear which one it is, and leave a comment that users don't have to change imports, and they can only add the call to patch on top of the file before they run imports.

We can also here show training times and compare them.

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:133]: I don't think it makes sense to link to a test file on another repo, we should link to a documentation page.

I didn't read the test file, but since the training procedures are different, the outputs can't be guaranteed to be the same. They can only be tested to be the same on a few toy examples, which doesn't guarantee equivalence.

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:145]: please use a tempfile rather than saving in the current directory.

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:176]: we shouldn't have empty cards from our own examples, please include a few basic info about the model and the data and the performance here.

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:207]: do we need to test this? I think this is well tested in other places

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:251]:
```suggestion
# in faster inference times, since loading a persisted model will always load the objects
# exactly as they were saved.
```

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:276]: need to pass the token as well.

File: skops/_min_dependencies.py
Comment (adrinjalali) [inline skops/_min_dependencies.py:14]: do we really need a very recent release here? These are minimum version dependencies.

**Review by adrinjalali — COMMENTED** [review]
Thanks for the PR @ahuber21 , it's in a pretty good shape.

I left a few minor comments. I haven't made sure the line lengths are correct. You don't need to "accept suggestion" here, you can take the texts and apply them the way you like.

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:67]: On my machine the bottleneck is up- and downloading the models. I'm optimistic all the code will run in <10s.

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:100]: I changed the imports. Training for kNN is <1s and in many cases `sklearnex` comes with a few percent of overhead. We're working on that, but for now I think it wouldn't help to add it.

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:134]: sklearnex is tested on the stock scikit-learn tests and is therefore "as tested as scikit-learn". Of course the implementations differ and it is impossible to reproduce the computations exactly (which isn't the goal in the first place). But all algorithms are required to produce results that are in very good agreement (typically the diff is required to be <1E-5).
We can definitely discuss how this is phrased, but the overall message should be that it is save to use sklearnex as a drop-in replacement and that statistically equivalent predictions will be made from sklearnex algos.

File: examples/plot_intelex.py
Comment (napetrov) [inline examples/plot_intelex.py:134]: Think that keeping this simple would make sense 
```suggestion
# We see that inference using ``sklearnex`` is a lot faster while achieving the
# same log-loss score.
```

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py:208]: Wanted to demonstrate that we can download and use optimized models. Should I remove it?

File: skops/_min_dependencies.py
Comment (ahuber21) [reply skops/_min_dependencies.py:14]: tbh: I don't know, so I added what I tested. We can go with a 1y old release. As far as I can tell this should work too. But I wouldn't want to put an even older release there.

File: examples/plot_intelex.py
Comment (adrinjalali) [reply examples/plot_intelex.py]: in that case, it does help to add it, since right now we're only selectively showing data, not the whole truth, and that's not what we wanna do here.

File: examples/plot_intelex.py
Comment (adrinjalali) [reply examples/plot_intelex.py:241]: I don't mind, as you wish. I'm not sure if it's adding to the message here though.

**Comment by adrinjalali** [comment]
Build failing, once you're done with edits, please ping me for a second review.

**Comment by ahuber21** [comment]
Don't quite understand what happened with the ubuntu tests. To me it looked unrelated. Anyway, I just pushed an update without the "Download model and re-evaluate" section, so there'll be a new run.
From my side, I've added everything. Let me know what you think

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py]: removed

File: examples/plot_intelex.py
Comment (ahuber21) [reply examples/plot_intelex.py]: I have added a `perf_counter` for the fit stage for completeness. On my machine, both fit times were ~0.05s. The numbers printed, but doing anything else with them (like calculating ratios) won't help if we don't do multiple runs. I hope you're fine with how it is now. Let me know :)

File: examples/plot_intelex.py
Comment (adrinjalali) [inline examples/plot_intelex.py:109]:
```suggestion
# We are not comparing the k-NN fit times, since this is not a compute-intensive
# task and both are typically very fast.
```

File: docs/examples.rst
Comment (adrinjalali) [inline docs/examples.rst:20]:
```suggestion
- Using Intel(R) Extension for scikit-learn:
    :ref:`sphx_glr_auto_examples_plot_intelex.py` is an example of using
    Intel(R) Extension for scikit-learn to speed up inference of classical
    machine learning algorithms and how performance-optimized models work with
    Hugging Face Hub.
```

(just wrapping lines)

**Review by adrinjalali — COMMENTED** [review]
Other than the nits, LGTM. WDYT @BenjaminBossan

**Comment by ahuber21** [comment]
Force-pushed again because in the intro I still said that we download the models again. Nothing else changed.

**Comment by adrinjalali** [comment]
@ahuber21 we squash and merge anyway, please avoid force pushing since it makes reviewing changes harder :)

**Review by BenjaminBossan — APPROVED** [review]
Yes, looks good, thanks Andreas

**Comment by ahuber21** [comment]
> I think this will depend on how exactly the patching works, which I don't know. Ideally, there would be an easy way to check on each estimator if it was patched, something like hasattr(estimator, "_uses_intelex") or sklearnex.check_is_patched(estimator), I believe this would be really useful.

@BenjaminBossan this feature is now added in the latest release.

```python
from sklearnex import is_patched_instance
my_instance = pickle.load(open("some_file.pkl", "rb"))
is_patched_instance(my_instance)  # true if it was trained using sklearnex, false otherwise 
```

**Pull request merged**
